File: src/main/scala/jk_5/nailed/server/chat/ChatComponentConverter.scala
/*
 * Nailed, a Minecraft PvP server framework
 * Copyright (C) jk-5 <http://github.com/jk-5/>
 * Copyright (C) Nailed team and contributors <http://github.com/nailed/>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the MIT License.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the MIT License along with
 * this program. If not, see <http://opensource.org/licenses/MIT/>.
 */

package jk_5.nailed.server.chat

import jk_5.nailed.api.chat._
import net.minecraft.event
import net.minecraft.util._
import org.apache.logging.log4j.LogManager

import scala.collection.convert.wrapAsScala._

/**
 * No description given
 *
 * @author jk-5
 */
object ChatComponentConverter {

  private val logger = LogManager.getLogger

  implicit def toVanilla(component: BaseComponent): IChatComponent = {
    val base = singleComponentToVanilla(component)
    if(component.getExtra != null){
      for(child <- component.getExtra){
        base.appendSibling(toVanilla(child))
      }
    }
    base
  }

  implicit def arrayToVanilla(comp: Array[BaseComponent]): IChatComponent = {
    val base = new ChatComponentText("")
    for(c <- comp){
      base.appendSibling(c)
    }
    base
  }

  def singleComponentToVanilla(component: BaseComponent): IChatComponent = {
    val base = component match {
      case t: TextComponent => new ChatComponentText(t.getText)
      case t: TranslatableComponent => new ChatComponentTranslation(t.getTranslate, t.getWith)
      case t =>
        logger.warn("Was not able to convert component {0} to vanilla", t.toString)
        return null
    }
    val style = new ChatStyle
    style.setColor(component.getColorRaw)
    style.setBold(component.isBoldRaw)
    style.setItalic(component.isBoldRaw)
    style.setUnderlined(component.isBoldRaw)
    style.setStrikethrough(component.isBoldRaw)
    style.setObfuscated(component.isObfuscatedRaw)
    if(component.getHoverEvent != null){
      val e = component.getHoverEvent
      val newAction = e.getAction match {
        case HoverEvent.Action.SHOW_ACHIEVEMENT => event.HoverEvent.Action.SHOW_ACHIEVEMENT
        case HoverEvent.Action.SHOW_ITEM => event.HoverEvent.Action.SHOW_ITEM
        case HoverEvent.Action.SHOW_TEXT => event.HoverEvent.Action.SHOW_TEXT
        case _ => null
      }
      if(newAction != null){
        style.setChatHoverEvent(new event.HoverEvent(newAction, e.getValue))
      }
    }
    if(component.getClickEvent != null){
      val e = component.getClickEvent
      val newAction = e.getAction match {
        case ClickEvent.Action.OPEN_FILE => event.ClickEvent.Action.OPEN_FILE
        case ClickEvent.Action.OPEN_URL => event.ClickEvent.Action.OPEN_URL
        case ClickEvent.Action.RUN_COMMAND => event.ClickEvent.Action.RUN_COMMAND
        case ClickEvent.Action.SUGGEST_COMMAND => event.ClickEvent.Action.SUGGEST_COMMAND
        case _ => null
      }
      if(newAction != null){
        style.setChatClickEvent(new event.ClickEvent(newAction, e.getValue))
      }
    }
    base.setChatStyle(style)
    base
  }

  @inline implicit def convertColor(color: ChatColor): EnumChatFormatting = if(color == null) null else EnumChatFormatting.values().find(c => c.toString() == "\u00a7" + color.getCode).orNull
}

File: src/main/java/jk_5/nailed/server/chat/ChatComponentConverter.java
package jk_5.nailed.server.chat;

import jk_5.nailed.api.chat.*;
import net.minecraft.util.*;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

@Deprecated
public class ChatComponentConverter {

    private static final Logger logger = LogManager.getLogger();

    public static IChatComponent toVanilla(BaseComponent component){
        IChatComponent base = singleComponentToVanilla(component);
        if(component.getExtra() != null){
            for(BaseComponent child : component.getExtra()){
                base.appendSibling(toVanilla(child));
            }
        }
        return base;
    }

    public static IChatComponent arrayToVanilla(BaseComponent[] comp){
        IChatComponent base = new ChatComponentText("");
        for(BaseComponent c : comp){
            base.appendSibling(toVanilla(c));
        }
        return base;
    }

    public static IChatComponent singleComponentToVanilla(BaseComponent component){
        IChatComponent base;
        if(component instanceof TextComponent){
            base = new ChatComponentText(((TextComponent) component).getText());
        }else if(component instanceof TranslatableComponent){
            base = new ChatComponentTranslation(((TranslatableComponent) component).getTranslate(), ((TranslatableComponent) component).getWith());
        }else{
            logger.warn("Was not able to convert component {0} to vanilla", component.toString());
            base = null;
        }
        ChatStyle style = new ChatStyle();
        style.setColor(convertColor(component.getColorRaw()));
        style.setBold(component.isBoldRaw());
        style.setItalic(component.isBoldRaw());
        style.setUnderlined(component.isBoldRaw());
        style.setStrikethrough(component.isBoldRaw());
        style.setObfuscated(component.isObfuscatedRaw());
        if(component.getHoverEvent() != null){
            HoverEvent e = component.getHoverEvent();
            net.minecraft.event.HoverEvent.Action newAction;
            switch(e.getAction()){
                case SHOW_ACHIEVEMENT:
                    newAction = net.minecraft.event.HoverEvent.Action.SHOW_ACHIEVEMENT;
                    break;
                case SHOW_ITEM:
                    newAction = net.minecraft.event.HoverEvent.Action.SHOW_ITEM;
                    break;
                case SHOW_TEXT:
                    newAction = net.minecraft.event.HoverEvent.Action.SHOW_TEXT;
                    break;
                default:
                    newAction = null;
                    break;
            }
            if(newAction != null){
                style.setChatHoverEvent(new net.minecraft.event.HoverEvent(newAction, arrayToVanilla(e.getValue())));
            }
        }
        if(component.getClickEvent() != null){
            ClickEvent e = component.getClickEvent();
            net.minecraft.event.ClickEvent.Action newAction;
            switch(e.getAction()){
                case OPEN_FILE:
                    newAction = net.minecraft.event.ClickEvent.Action.OPEN_FILE;
                    break;
                case OPEN_URL:
                    newAction = net.minecraft.event.ClickEvent.Action.OPEN_URL;
                    break;
                case RUN_COMMAND:
                    newAction = net.minecraft.event.ClickEvent.Action.RUN_COMMAND;
                    break;
                case SUGGEST_COMMAND:
                    newAction = net.minecraft.event.ClickEvent.Action.SUGGEST_COMMAND;
                    break;
                default:
                    newAction = null;
                    break;
            }
            if(newAction != null){
                style.setChatClickEvent(new net.minecraft.event.ClickEvent(newAction, e.getValue()));
            }
        }
        base.setChatStyle(style);
        return base;
    }

    public static EnumChatFormatting convertColor(ChatColor color){
        if(color == null){
            return null;
        }else{
            for(EnumChatFormatting c : EnumChatFormatting.values()){
                if(c.toString().equals("\u00a7" + color.getCode())){
                    return c;
                }
            }
            return null;
        }
    }
}
